### PR TITLE
Preprocess cabal-install.cabal: don't leak dev-details into production

### PR DESCRIPTION
**Pull request by phadej**
I used `m4`, but if the approach accepted generally, I can rewrite it in Haskell-only solution.

In short:
- for production the generated `cabal-install.cabal` is `cabal-version: >=1.10` and is very plain
- for development/CI we use `cabal-version: 2.0` so we can use internal libraries!

In the long run - when our support window allows - the latter version can become the default.

The only drawback, is that one have to be careful not to commit development `cabal-install.cabal` file.

### CODE REVIEW

**Comment by 23Skidoo** [comment]
> The only drawback, is that one have to be careful not to commit development `cabal-install.cabal` file.

Perhaps we should add `cabal-install.cabal` to `.gitignore` then? It'll still be possible to commit it with with `git add --force`. And there should obviously be a CI meta check that the generated production `cabal-install.cabal` file is identical to the one in the repo.

**Comment by hvr** [comment]
@23Skidoo otoh, if you can have that metacheck, why would the generated .cabal file ever diverge from the one you'd generate automatically? i.e. when would you ever modify your cabal-install.cabal file w/o having modified the master one?

**Comment by 23Skidoo** [comment]
idk, people do strange things sometimes.

**Comment by 23Skidoo** [comment]
I think it's a good idea to have such checks for all generated files that are also committed to the repo.

**Comment by hvr** [comment]
I'm all for the check; I was just wondering whether it's still needed to gitignore if you have that check

**Comment by 23Skidoo** [comment]
If it's in `.gitignore`, it doesn't clutter `git status` output and you can still run `git commit -a`

**Comment by phadej** [comment]
I think if it's committed, than `.gitignore` doesn't apply to that file. `git` already knows about it, so to say.

**Comment by 23Skidoo** [comment]
Hmm, yes, you're right, for some reason I was sure that this worked.

**Comment by 23Skidoo** [comment]
`git update-index --assume-unchanged cabal-install/cabal-install.cabal` seems to work, however.

**Comment by phadej** [comment]
Rewrote using DIY-preprocessor, but it's way simpler syntax in comparison to m4 (we don't need all that power). I can also split `cabal-lib-solver`, but then I'll need to move the files into separate folders.

**Comment by phadej** [comment]
CI, was green, so I squased everything, and IMO it's mergeable.

I'm not sure if we have consensus on what to do regarding `git` and `cabal-install.cabal`, in any case that can be changed if there's some good way.

File: cabal-install/cabal-install.cabal.pp [hunk 1]
```diff
@@ -0,0 +1,642 @@
...
+#
+# Package Description
+#
+Synopsis:           The command-line interface for Cabal and Hackage.
+Description:
+    The \'cabal\' command-line program simplifies the process of managing
+    Haskell software by automating the fetching, configuration, compilation
+    and installation of Haskell libraries and programs.
+    .
+    This package only provides an executable and cannot be used as a
```
Comment (quasicomputational) [inline cabal-install/cabal-install.cabal.pp]: Is this caveat needed any more?

Comment (phadej) [reply cabal-install/cabal-install.cabal.pp]: I think not.

**Comment by 23Skidoo** [comment]
We can make `make cabal-install-[dev,monolithic]` print a message telling the user to run `git update-index --assume-unchanged cabal-install/cabal-install.cabal` to make Git ignore changes to that file. I don't think it should do that by default.

**Comment by phadej** [comment]
@23Skidoo good idea, I'll add that.

File: travis-script.sh [hunk 2]
```diff
@@ -165,17 +167,20 @@ fi
 
 # Haddock
 # TODO: Figure out why this needs to be run before big tests
-if [ "$TRAVIS_OS_NAME" = "linux" ]; then
-    (cd cabal-install && timed ${CABAL_INSTALL_SETUP} haddock --builddir=${CABAL_INSTALL_BDIR} ) || exit $?
-fi
+# TODO: cabal-install doesn't have public lib, haddock doesn't make sense
+#if [ "$TRAVIS_OS_NAME" = "linux" ]; then
+#    timed cabal new-haddock cabal-install
```
[the remark this answers is not in the document]
Comment (phadej) [reply travis-script.sh]: I think `haddock` have to learn about (public) internal libraries properly. Or should it handle them already somehow, @alexbiehl , @hvr?

[the remark this answers is not in the document]
Comment (phadej) [reply travis-script.sh]: i.e. I'd leave this out (and create a tracking issue),

[the remark this answers is not in the document]
Comment (hvr) [reply travis-script.sh]: afaik, haddock gets fed internal libraries as if they were ordinary libraries by cabal; the thing we might need to adapt is to create a "TOC" maybe... but haven't thought much about it yet;

**Comment by phadej** [comment]
I'll forward patch this to `master` tomorrow (on Monday)

**Pull request merged**